src/pp.rs

```diff
@@ -373,14 +368,47 @@ impl<'a> DirectiveProcessor<'a> {
         Ok(())
     }
 
-    fn parse_if_directive(&mut self, directive_location: Location) -> Step<()> {
-        self.parse_if_like_directive(directive_location, |this, location| {
-            if let LexerTokenValue::Integer(i) = this.expect_a_lexer_token(location)?.value {
+    fn evaluate_if_expression(&mut self, location: Location, line: Vec<Token>) -> Step<bool> {
+        let line: Vec<_> = line.into_iter().map(|t| t.value).collect();
+
+        match *line {
+            [TokenValue::Integer(ref i)] => {
+                // Evaluate "#if 123"
                 Ok(i.value != 0)
-            } else {
+            }
+            [TokenValue::Ident(ref ident), TokenValue::Punct(Punct::LeftParen), 
+                TokenValue::Ident(ref ident2), TokenValue::Punct(Punct::RightParen)] |
+            [TokenValue::Ident(ref ident), TokenValue::Ident(ref ident2)] if ident == "defined" => {
+                // Evaluate "#if defined XYZ" and "#if defined(XYZ)"
+                Ok(self.defines.contains_key(ident2))
+            }
+            [TokenValue::Punct(Punct::Bang), TokenValue::Ident(ref ident), TokenValue::Punct(Punct::LeftParen), 
+                TokenValue::Ident(ref ident2), TokenValue::Punct(Punct::RightParen)] |
+            [TokenValue::Punct(Punct::Bang), TokenValue::Ident(ref ident), TokenValue::Ident(ref ident2)] if ident == "defined" => {
+                // Evaluate "#if !defined XYZ" and "#if !defined(XYZ)"
+                Ok(!self.defines.contains_key(ident2))
+            }
+            [TokenValue::Ident(ref ident)] => {
+                // Handle "#if XYZ"
+                match self.defines.get(ident) {
+                    Some(d) => {
+                        let line = d.tokens.clone();
+                        self.evaluate_if_expression(location, line)
```

Comment:
Technically this could allow for:

```c
#define XYZ !defined(A)
#if XYZ
```

Which wouldn't work in the C/GLSL preprocessor because the `defined` token needs to be pre macro expansion. Maybe we could only do the recursive call is `line` is of length 1 to avoid this case? It would still pass the tests you added.